0013-secure-http-api.adoc

```diff
@@ -0,0 +1,152 @@
+
+The topic of authentication is not so clear cut.
+Authentication is actually out of scope for this spike but briefly considering here for completeness.
+
+.Client authentication refers to three things:
+* Identity – who claims to be making an API request?
+* Authentication – are they really who they say they are?
+* Authorization – are they allowed to do what they are trying to do?
+
+At the moment only the first two are of interest to us, all users are allowed to do anything.
+
+.Considerations:
+* API key
+* username/password
+* Basic authentication with TLS
+* OAuth
+** Oauth1.Oa
+** OAuth2
+
+Recommend to leave authentication out of scope and only consider encryption.
+
+=== Recap on TLS
+
+The industry standard for encrypting traffic over a network is Transport Layer Security (TLS).
+
+.What TLS gives us and whether we need it:
+* Authentication of cnd - needed, allows client to verify it is connecting to the correct cnd.
+* Data confidentiality - needed, protects against attacker learning the secret.
+* Date integrity - needed, protects against modification of the data i.e. changing the redeem address.
+* Replay protection - not needed, replay attack vector currently non-existent for cnd.
+* Sequence protection - no obvious need.
+
+=== Certificates
+
+TLS requires a certificate (private/public keypair) to be preset on the server i.e., cnd.
+Note, this negatively impacts our quarterly OKR #48 (Make COMIT easily accessible).
+
+There are multiple solutions to getting a certificate onto the machine running cnd, each with different pros/cons.
+
+==== Signed certificate by trusted CA
+
+Use case: running cnd on an internet facing host with a registered domain.
+Purchase a certificate signed by a trusted CA.
+
+.pros
+* This is how it would need to be done if cnd is offered as a service i.e., to mobile clients
+* Industry standard
+
+.cons
+* Resource intensive, money and time.
+* Unnecessary for use cases where cnd is on the same machine as the client
+* Not suitable for development
+
+==== Signed certificate by private CA
+
+Use case: connecting to `comit-i` with the machine running cnd.
+Create a private CA and configure the host machine to trust certificates signed by the private CA.
+
+.pros
+* Free
+* Makes using the client with HTTPS seamless
+
+.cons
+* Requires root access to the host machine
+* Complex if done manually, however there is a tool (untested) that claims to simplify this: https://blog.filippo.io/mkcert-valid-https-certificates-for-localhost/
+* Operator of cnd still has to create and sign certificates for use by cnd i.e., increases complexity
+
+=== Self signed certificate
+
+Use case: connecting to `comit-i` with the machine running cnd.
+Create a self signed certificate.
+
+.pros
+* Free
+* Within the realm of typical developer knowledge
+
+.cons
+* Adds complexity to the set up of cnd
```

Comment:
Why not:
- cnd generates certificates when starting up if not yet existing and if `webgui` is activated and `webgui.tls` (new option) is activated. This will be a CA, Public Cert and Private Cert
- User is told to install the freshly generated CA

That's it.

Feel free to at `lnd` as it does generate certificates for its gRPC API on the spot as described above (it does not generate a CA though, only a `tls.cert` and `tls.key`).

Reply:
I think I get what you are saying but am confused by the terminology - I'm only just across all the TLS certificate terms as is.  Either way we still either make the user click through security warnings or require root to install the CA.

Reply:
I do not think that `cnd` should create the certificate.  I think we should stick to the UNIX principle of doing one thing.  The user can generate the certificate.